x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -722,37 +724,56 @@ public void refreshToken(String refreshToken, ActionListener<Tuple<UserToken, St
 
     /**
      * Performs an asynchronous search request for the token document that contains the {@code refreshToken} and calls the listener with the
-     * {@link SearchResponse}. In case of recoverable errors the SearchRequest is retried using an exponential backoff policy.
+     * {@link SearchResponse}. In case of recoverable errors the {@code SearchRequest} is retried using an exponential backoff policy.
      */
     private void findTokenFromRefreshToken(String refreshToken, Iterator<TimeValue> backoff, ActionListener<SearchHit> listener) {
+        Tuple<Version, String> versionAndRefreshTokenTuple;
+        try {
+            versionAndRefreshTokenTuple = unpackVersionAndPayload(refreshToken);
+        } catch (IOException e) {
```

Comment:
> (a) I think unpackVersionAndPayload should handle the old-token case rather than all callers having to deal with it.

Agreed! Will do, I found a bug on it as well.

> (b) I think it would be better if we could distinguish between old bas64-UUID tokens and new base64-data-stream tokens without relying on an IOException. Is the length not enough? Do we need to prefix the new tokens with a non-base64 character?

Okay. I will use the length. Refresh tokens contain `-` which is not part of the base64 alphabet in RFC4648 so it would throw reliably, but I get that throwing for code flow is undesirable.

Reply:
Though, moving to hashes will change lengths ... and the alphabet... I understand your concern, but since we have a "data packet" that changed format, and it was not versioned, the try-parse-header-else-assume-unversioned does not sound too terrible.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/40742/commits/3952d8abb380acadc481d3f5169473ad1d0a62dc .
I've gave it some thought and I believe trying to parse the version is superior to other forms "recognizing" the old format. Relying on heuristics like length and alphabet is just continuing under the "unversioned way" paradigm. I think they are brittle because alphabets are more-or-less the same base64 and lengths would change when we'll store hashes for doc ids.  But indeed, this trial-and-error should not throw to callers since it is to be expected, and should be handled inside a function.